pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,202 @@
+)
+
+type ObjectType string
+
+const (
+	NetworkPolicy ObjectType = "np"
+	Pod           ObjectType = "pod"
+)
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    EventType
+	ObjType ObjectType
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
+	}()
+}
+
+func (rw *RemoteCluster) GetPods() []interface{} {
+	return rw.podInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rw.networkPolicyInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan *Event) {
+	rw.eventChanMutex.Lock()
+	rw.eventChan = eventChan
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) enqueueEvent(event *Event) {
+
+	// lock used as memory barrier to make sure eventChannel ref
+	// is synchronized between threads
+	rw.eventChanMutex.Lock()
+	if rw.eventChan != nil {
+		rw.eventChan <- event
+	}
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) OnAdd(obj interface{}) {
+
+	var objType ObjectType
+
+	if pod, ok := obj.(*v1.Pod); ok {
+		klog.Infof("Pod discovered in %s: %s", rw.ClusterID, pod.GetSelfLink())
+		objType = Pod
+	} else if np, ok := obj.(*v1net.NetworkPolicy); ok {
+		klog.Infof("NetworkPolicy discovered in %s: %s", rw.ClusterID, np.GetSelfLink())
+		objType = NetworkPolicy
+	} else {
```

Comment:
Defensive programming, when you are not sure what following coders will the codebase it's better to be ready. At least I share that defensiveness with @tpantelis  :)

Reply:
UT FTW. Have the behavior coded.

I disagree with that. That's what tests and reviews are for.

Reply:
But we depend on external libraries that will be feeding those events to us. It's why I feel the need to defend. It's not adding unecessary complexity to the code (2 lines aren't) can I sleep better? :D ;)

Reply:
You never know what strange behavior you might get in production so it doesn't hurt to trap and log it. It's similar to checking for `DeletedFinalStateUnknown`, eg

```
myObj, ok := obj.(*MyObject)
if !ok {
    tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
    if !ok {
        klog.Errorf(...)
        return
    }
    myObj, ok = tombstone.Obj.(*MyObject)
    if !ok {
        klog.Errorf("...)
        return
    }
}
```

Can `obj` really be anything but a `MyObject` or a `DeletedFinalStateUnknown` and/or can `tombstone` really be anything but a `MyObject`?  I wouldn't think so yet every example I've seen does this same defensive programming...

Reply:
It's not exactly the same as the DeletedFinalStateUnknown.

I don't agree with the idea of this but not going to make a fuzz about it.

Reply:
> It's not exactly the same as the DeletedFinalStateUnknown.
> 

Same concept, ie verifying casts are what you expect from an informer event and logging otherwise.  You seem to agree that the checking is necessary for `DeletedFinalStateUnknown` but not in this case, that seems inconsistent to me...

Reply:
I can clarify so its not inconsistent any more, since I don't want to get misinterpreted. 

The `DeletedFinalStateUnknown` is documented as a valid input your handler can get in the event of a deletion event happening before the actual deletion occurs. As such it's part of the contract you have with the caller about possible inputs you are getting. 

In this case, however, you are subscribed to receive events of about those specific resources. And, in contrast to the delete event handler, there is no documentation saying that you will receive an object that is not going to be cast as the resource you expect to receive. Moreover, if you see usages inside Kubernetes and many other related projects were they make use of Informers (and know this libraries very well), they don't have any handling that would indicate the behavior that would warrant this defensive coding. As such, I think it is fair to assume that in fact the objects received during the addition and update events do in fact cast to the expected type.

Having explained this, I think it should now be clear that it's not an inconsistent criteria and you can now understand why I ask the question in the first place. If not, I would be happy to have this healthy discussion going (without getting way in the patch)

Reply:
I realize that `DeletedFinalStateUnknown` is documented as a valid input - my analogy wrt `DeletedFinalStateUnknown` is the defensive coding surrounding `DeletedFinalStateUnknown` and its contents. Following your argument, then the only other possible input will be `DeletedFinalStateUnknown` and the only possible content of `DeletedFinalStateUnknown` will be an instance your resource type. If that is the case, then why don't other usages in other projects who know the libraries very well simply do:

```
myObj, ok := obj.(*MyObject)
if !ok {
    myObj = obj.(cache.DeletedFinalStateUnknown).Obj.(*MyObject)
}
```
In reality, most likely the above code **is** fine and no defensive coding wrt casting is really needed, yet... when you do see such defensive coding in certain cases out there in the wild it leaves just a little bit of doubt in your mind. Just sayin...

Reply:
Since you're going to be handling based on the type anyway later [1], I see no harm in just letting the objType be nil (or whatever default it is) and enqueueing that..

In fact this whole type checking/casting logic can (and probably should) sit peacefully on that side of the queue and you can handle unknown types there in a centralized way (instead of the copy-paste-fest you have in this file ;))